### PR TITLE
Mongodb

### PR DESCRIPTION
**Pull request by omercotkd**

### CODE REVIEW

File: libs/vector_storage/vector_provider/mongodb.py [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+            [
+                {
+                    "$vectorSearch": {
+                        "index": "default_vector_index",
+                        "path": "values",
+                        "filter": (
+                            {"$and": builded_filters} if builded_filters else None
+                        ),
+                        "queryVector": query_vector,
+                        "numCandidates": k * 10,
```
Comment (liranzxc) [inline libs/vector_storage/vector_provider/mongodb.py:66]: Why num of candidates is k * 10 ? Not just k .
Also why we need limit k ?

File: libs/vector_storage/vector_provider/mongodb.py [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+                        "limit": k,
+                    }
+                },
+                {"$addFields": {"score": {"$meta": "vectorSearchScore"}}},
+                # {"$match": {"score": {"$gte": threshold}}},
+                {"$sort": {"score": -1}},
```
Comment (liranzxc) [inline libs/vector_storage/vector_provider/mongodb.py:72]: We should filter after Ann by threshold after projection the score

File: libs/vector_storage/vector_provider/mongodb.py [hunk 3]
```diff
@@ -0,0 +1,107 @@
+from abc import ABC
+from typing import List, Optional
+
+from pymongo import MongoClient
```
Comment (liranzxc) [inline libs/vector_storage/vector_provider/mongodb.py:4]: I think we should have schema of mongodb , we should have index of Metadata fields

**Review by liranzxc — CHANGES_REQUESTED** [review]
Great job

File: libs/vector_storage/vector_provider/mongodb.py [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+        items = []
+        for doc in documents:
+            vector = self.generate_vector(doc)
+            _id = self.generate_id(doc)
+            record: EmbeddingDocument = EmbeddingDocument(
+                id=_id, values=vector, metadata=doc
+            )
+            items.append(record.model_dump())
+
+        self.document_collection.insert_many(items)
```
Comment (liranzxc) [inline libs/vector_storage/vector_provider/mongodb.py:96]: i think we should do upsert , in case the id is same id , we don't want append new data

**Review by liranzxc — CHANGES_REQUESTED** [review]

File: libs/vector_storage/vector_provider/mongodb.py (hunk 1 above)
Comment (omercotkd) [reply libs/vector_storage/vector_provider/mongodb.py:66]: image
https://www.mongodb.com/docs/atlas/atlas-vector-search/vector-search-stage/#mongodb-pipeline-pipe.-vectorSearch

The docs of the "vectorSearch" pipeline should answer those questions

File: libs/vector_storage/vector_provider/mongodb.py (hunk 2 above)
Comment (omercotkd) [reply libs/vector_storage/vector_provider/mongodb.py:72]: Yes we should but the score is different for some reason its two decimal points, e.g.: 0.09 and not 0.9.
should I just multiply the threshold by 0.1 to normalize the threshold?

File: libs/vector_storage/vector_provider/mongodb.py (hunk 4 above)
Comment (omercotkd) [reply libs/vector_storage/vector_provider/mongodb.py:96]: I would add an index to make the id unique, do you still want to upsert so the data will be updated if needed?

Comment (liranzxc) [reply libs/vector_storage/vector_provider/mongodb.py:96]: Yes , just not create duplicate record

File: libs/vector_storage/vector_provider/mongodb.py (hunk 2 above)
Comment (liranzxc) [reply libs/vector_storage/vector_provider/mongodb.py:72]: According to docs that value should be between 0 to 1

**Review by liranzxc — APPROVED** [review]

**Pull request merged**